Replace running stitch algorithm to give consistent stitch lengths

george-steel opened the pull request:
Use a modified Zhao-Saalfeld algorithm for running stitch instead of the previous Douglas-Peucker + additional subdivision. This makes stitch lengths much more consistent while still shortening stitches in order to stay within curve tolerance.

* Improves runtime of running stitch from quadratic to linear time.
* Also fix bug with interaction of contour and center-walk underlay in satin stitches.

Fixes #1676

george-steel commented:
![20230122_145744](https://user-images.githubusercontent.com/633295/213937497-375e6220-63dc-488d-8c4d-ac66b7465aa7.jpg)

Here are some before and after shots illustrating the effect of the patch. The spiral ended up with quite an inconsistent length under the old algorithm while the new one has the stitch length smoothly decreasing as the curve tightens. Contour fill also looks a fair bit cleaner although it still has a strong moire effect, which I hope to reduce in another randomization patch.

george-steel commented:
@tyrosinase Hopefully this makes running stitch tolerance useful for you.

tyrosinase commented:
> @tyrosinase Hopefully this makes running stitch tolerance useful for you.

That looks excellent!

lexelby commented:
I got this while testing:

```
Traceback (most recent call last):
  File "inkstitch.py", line 84, in <module>
    extension.run(args=remaining_args)
  File "/home/lex/repos/inkstitch/src/inkex/inkex/base.py", line 231, in run
    self.save_raw(self.effect())
  File "/home/lex/repos/inkstitch/lib/extensions/stitch_plan_preview.py", line 39, in effect
    patches = self.elements_to_stitch_groups(self.elements)
  File "/home/lex/repos/inkstitch/lib/extensions/base.py", line 202, in elements_to_stitch_groups
    patches.extend(element.embroider(last_patch))
  File "/home/lex/repos/inkstitch/lib/elements/element.py", line 398, in embroider
    patches = self.to_stitch_groups(last_patch)
  File "/home/lex/repos/inkstitch/lib/elements/stroke.py", line 477, in to_stitch_groups
    patch = self.running_stitch(path, self.running_stitch_length, self.running_stitch_tolerance)
  File "/home/lex/repos/inkstitch/lib/elements/stroke.py", line 435, in running_stitch
    stitches = running_stitch(path, stitch_length, tolerance)
  File "/home/lex/repos/inkstitch/lib/stitches/running_stitch.py", line 222, in running_stitch
    stitches.extend(stitch_curve_even(curve, stitch_length, tolerance))
  File "/home/lex/repos/inkstitch/lib/stitches/running_stitch.py", line 195, in stitch_curve_even
    stitch, newidx = take_stitch(last, points, i, stitch_len, tolerance)
  File "/home/lex/repos/inkstitch/lib/stitches/running_stitch.py", line 174, in take_stitch
    cut = sleeve.cutSegment(start, last, p)
  File "/home/lex/repos/inkstitch/lib/stitches/running_stitch.py", line 121, in cutSegment
    if segArc.containsAngle(self.a):
AttributeError: 'NoneType' object has no attribute 'containsAngle'

```

on path: 

```svg
    <path
       style="fill:none;fill-opacity:0.582241;stroke:#000000;stroke-width:2.51339;stroke-dasharray:30.1606, 2.51339;stroke-dashoffset:0;stroke-opacity:1;stop-color:#000000"
       d="m 107.94542,251.35155 h 33.07957 v -16.81067 16.8606 l 32.56762,0"
       id="path12093" />
```

lexelby commented:
I came up with the above test while investigating whether this implementation is vulnerable to this purported bug in the zhao-saalfeld algorithm itself: https://github.com/mapnik/mapnik/issues/2748#issuecomment-145349315

I think it is vulnerable.  Here's an example:

![image](https://user-images.githubusercontent.com/4446653/213941837-b2a3026a-8f6d-4ebb-8fb8-ca6ee3b1f135.png)

This is a path that backtracks for about half a millimeter.  I've run Stitch Plan Preview with "lower opacity" enabled, so the grey dotted line above is the original path, and the solid black line is the produced stitch plan.  The backtrack is removed, even when I reduce the running stitch tolerance to 0.01.  Can we make use of the fix hinted at by the commenter?

lexelby commented:
As to that moire effect you noted, I think that may be rather more desirable than not, at least for some use cases.  It looks nice!  In some ways those kinds of effects are natural, for example, the diagonal lines that emerge from normal fill stitch.  In any case, I agree a randomization sounds like a promising path to explore.

george-steel commented:
@lexelby I fixed the backtracking case (which was a missing case in segment cutting and a missing abs in path_to_curves. With the fix, this is not vulnerable to the ZS bug you linked to.

lexelby commented:
Wow, that was fast!!  From what I can see this trend to produce running stitch with much more even stitch lengths.

What's the running time like?  How does the running time compare to the old algo?

george-steel commented:
> What's the running time like? How does the running time compare to the old algo?

On my test it ran quite quickly. It also has the advantage of running in linear time while the old algorithm had n**1.5 average case and n**2 worst case performance.

lexelby commented:
Verified: my test now no longer crashes, and the backtrack section is not eliminated.

My concern about running time had to do mostly with switching from an algorithm partly implemented in Shapely to a pure python approach.  Sometimes O(n**1.5) or O(n**2) can be better if your O(n) constant is big. :)

lexelby commented:
Just got this:

```
Traceback (most recent call last):
  File "/home/lex/repos/inkstitch/lib/elements/fill_stitch.py", line 563, in to_stitch_groups
    underlay_stitch_groups, start = self.do_underlay(underlay_shape, start)
  File "/home/lex/repos/inkstitch/lib/elements/fill_stitch.py", line 608, in do_underlay
    underpath=self.underlay_underpath))
  File "/home/lex/repos/inkstitch/lib/debug.py", line 133, in decorated
    result = func(*args, **kwargs)
  File "/home/lex/repos/inkstitch/lib/stitches/auto_fill.py", line 77, in auto_fill
    staggers, skip_last)
  File "/home/lex/repos/inkstitch/lib/debug.py", line 133, in decorated
    result = func(*args, **kwargs)
  File "/home/lex/repos/inkstitch/lib/stitches/auto_fill.py", line 630, in path_to_stitches
    stitches.extend(travel(travel_graph, edge[0], edge[1], running_stitch_length, running_stitch_tolerance, skip_last))
  File "/home/lex/repos/inkstitch/lib/stitches/auto_fill.py", line 593, in travel
    stitches = running_stitch(path, running_stitch_length, running_stitch_tolerance)
  File "/home/lex/repos/inkstitch/lib/stitches/running_stitch.py", line 232, in running_stitch
    for curve in path_to_curves(points):
  File "/home/lex/repos/inkstitch/lib/stitches/running_stitch.py", line 214, in path_to_curves
    last_seg = points[1] - points[0]
IndexError: list index out of range
```

for this:

```svg
<path
       style="fill:#000000;fill-opacity:1;stroke:none;stroke-width:0.245669;stroke-dasharray:0.245669, 0.245669;stroke-dashoffset:0;stroke-opacity:1;stop-color:#000000"
       d="m 261.78174,181.01883 c -13.41561,4.43243 -25.80675,9.79132 -31.06099,24.11291 -1.31888,3.5949 -1.89098,8.29657 0.43442,11.62588 7.04649,10.08859 29.259,8.08107 35.4052,17.6541 4.13585,6.44181 -3.07926,13.89585 -7.38514,17.86941 -4.65352,4.29437 -9.80345,8.21413 -14.98748,11.84118 -10.40049,7.27677 -29.96315,16.79704 -33.23311,30.57173 -2.43008,10.23669 9.51974,18.08793 16.94238,22.17527 26.83205,14.77529 56.39545,15.51415 84.92904,5.16706 10.0488,-3.64397 22.05165,-9.4595 28.88894,-18.08468 2.31591,-2.92151 3.97466,-6.54534 2.17206,-10.11882 -2.43196,-4.82114 -7.46815,-7.41919 -12.38095,-9.04235 -7.84107,-2.59064 -40.30442,-1.68729 -28.23727,-17.8694 7.27257,-9.75256 20.10574,-14.63708 30.40939,-20.23764 9.58526,-5.21008 26.13664,-17.42783 27.15122,-29.49525 1.23025,-14.63278 -23.05827,-11.4106 -31.71266,-11.4106 -2.60547,0 -7.9467,-0.19435 -7.38514,-4.09059 0.89803,-6.23068 9.79457,-12.89867 13.03263,-18.51527 5.15788,-8.9466 6.57328,-27.45448 -6.08186,-31.21763 -7.4921,-2.22786 -19.13074,11.49579 -24.76195,16.14706 -4.98883,4.12068 -11.57358,5.81292 -19.00586,7.31999 -8.75106,0.62365 -16.07863,3.26696 -23.13287,5.59764 z m 41.92152,17.00821 c 4.32329,4.28521 2.33203,8.95745 -1.73769,12.48707 -6.17852,5.35856 -16.84879,8.50114 -24.97913,6.8894 -13.01635,-2.58034 -18.58519,-13.42804 -5.21302,-20.88353 1.97972,-1.10376 3.98053,-2.15648 6.08187,-3.01411 3.0792,-1.25673 6.21206,-2.1531 9.55722,-1.50706 4.24633,3.00633 10.17364,-0.035 16.29075,6.02823 z m -24.2189,80.41231 c 2.07025,0.75353 2.42481,1.79376 3.36676,2.26059 5.78346,2.86627 11.15216,8.56633 5.86466,14.85527 -4.34221,5.16461 -14.22583,8.22206 -20.63493,8.82706 -9.54161,0.9007 -24.31155,-5.47791 -18.68005,-17.43882 1.49373,-3.1726 3.97011,-5.62556 7.16792,-7.10471 6.87352,-3.17935 14.62703,-4.69893 22.91564,-1.39939 z"
       id="path49235"
       sodipodi:nodetypes="sssssssssssssssssssscsssssscscsssssc" />
```

lexelby commented:
In my tests, this algorithm has comparable in running time to the old one.  It produces _way_ fewer stitches for the complex shape I tested.  That allowed me to use a significantly smaller tolerance and still use fewer stitches.  So far, I really like it!

It may be worth changing the default running stitch tolerance to 0.1mm.  It will produce nicer-looking results than the old algorithm, and it uses fewer stitches even with the tighter tolerance.

lexelby commented:
Your bounds check fix resolved that crash for me.

george-steel commented:
> It may be worth changing the default running stitch tolerance to 0.1mm. It will produce nicer-looking results than the old algorithm, and it uses fewer stitches even with the tighter tolerance.

I would agree. @kaalleen what do you think?

lexelby commented on lib/stitches/running_stitch.py:
```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
+        if diff == 0 or diff == math.pi:
+            return None
+        elif diff < math.pi:
+            return AngleInterval(angleA - 1e-6, angleB + 1e-6)
```
Why perturb the angles by `1e-6`?  A comment might be useful here.

george-steel replied:
added

lexelby commented on lib/stitches/running_stitch.py:
```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
+    def containsAngle(self, angle: float):
+        if self.all:
+            return True
+        return (angle - self.a) % tau <= (self.b - self.a) % tau
```
The paper devises a scheme where B is always greater than A (and B can be greater than `math.tau`).  That allows a simple comparison of `a <= angle <= b`.  Would that scheme work here?

george-steel replied:
II considered it 'd rather just leave a and b unrestricted and use modular comparison.

lexelby commented on lib/stitches/running_stitch.py:
```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
+            return other
+        elif other.all:
+            return self
+        elif self.containsAngle(other.a):
```
Similar here: the paper's scheme allows this to be simply`AngleInterval(max(self.a, other.a), min(self.b, other.b))`.

george-steel replied:
see response for the other spot. I wanted not to have restrictions on a and b.

lexelby replied:
@george-steel I don't think the paper restricts a and b, at least not in practical terms; it just requires that we add 360 degrees to b if b would be less than a.

george-steel replied:
I did it the other way at first, but the number of cases was becoming really messy and hard to keep track of. Modular comparison makes it a lot tidier.

lexelby commented on lib/stitches/running_stitch.py:
```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
         return []
+    distLeft = [0] * len(points)
+    for i in reversed(range(0, len(points) - 1)):
+        distLeft[i] = distLeft[i + 1] + points[i].distance(points[i+1])
```
I think it's possible to do this with numpy using just one or two function calls.  Might be faster?

george-steel replied:
Numpy doesn't take Point objects.

lexelby commented on lib/stitches/running_stitch.py:
```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
+        b = points[i + 1] - points[i]
+        aabb = (a * a) * (b * b)
+        abab = (a * b) * abs(a * b)
+        if aabb > 0 and 2 * abab < aabb:
```
How does this work?  I kind of vaguely recognize some formula here but I can't figure out what it is.

How would `aabb > 0` ever not be true? `a * a` is equivalent to the sum of the squares of a's X and Y components, so it must be positive.  Oh, or I guess it could be 0 if a is a zero-length vector.  Is that what you're testing for, whether either vector is of length zero?

george-steel replied:
`aabb == 0` if we have a duplicate point that would make either a or b equal to zero. The second part of the condition should technically take care of it, but there's floating point error.

lexelby commented:
I noticed that your algorithm brings back detection of corners.  I wonder if it also brings back any of the problems with that approach?

I tried removing the corner detection and now I understand why you're doing it.  I wonder if we can eliminate the original corner detection and instead add an angle check when `sleeve.contains(p - start)` is false.  If it's a sharp angle, choose `last` instead of cutting between `last` and `p`.

This would allow us to avoid doing the angle check along the entire line.  It also allows us to avoid stitching sharp corners that are extremely tiny, and thus really should be smoothed away by the running stitch algorithm.  That kind of corner can happen when tracing a bitmap, for example.

Take this path for example:

![image](https://user-images.githubusercontent.com/4446653/213970248-ebb5c29b-5a8c-4455-a3d3-12bb6e59215d.png)

There's a little zigzag there in the middle with a width of 0.2mm in total (so it deviates from the line by only 0.1mm in either direction).  Zoom out and it's barely visible.  I may not even know it's there.  If I set my running stitch tolerance to 0.2mm, 1mm, even 10mm, that stubborn zigzag still gets stitched.  It's gonna make tyrosinase's machine have a conniption.

If my idea works, that zigzag will all fall in the sleeve and get removed unless the tolerance is super low, which I would argue is what we want.  It's not just about the sharpness of the angle, it's how big the feature is that contains that angle, and Zhao-Saalfeld is the perfect algorithm to consider both of those qualities when making a decision about whether to stitch that corner.

lexelby commented:
All that said, I really like what you have so far, and the algorithm is so clean and simple!

george-steel commented:
@lexelby The corner letection is necessary for the stitch length adjustment, ensuring that polygons get evenly-split sides. If you try a contour fill without the corner detection you'll notice a huge difference. Unintentional corners are in interesting problem. Perhaps the best way to deal with those is thekeep the running stitch simple and manually apply simplification to autotraced curves.

lexelby commented:
I really don't want to go the route of requiring more work from the user.  Many of them simply won't know to do the simplification step.  I frequently don't _want_ to simplify my paths.

Maybe we can do a two-pass Zhao-Saalfeld?  The first pass is just to detect the sharp corners that actually matter and split the curve on them.  Then the second pass is how you have it now, the modified Zhao-Saalfeld with stitch length.  Would that work?

george-steel commented:
@lexelby I added a minimum segment length to the `path_to_curves` corner detection. This prevents micro-messes such as your zigzag example from causing way too many stitches, but still allows for corners of longer segments to be stitched exactly.

The specific minimum length of 2x the tolerance was set because it is approximately the minimum length of zigzag segments before the tolerance would force a corner anyways. Setting the minimum larger than this would break use of the running stitch for custom decorative stitches using pattern along path.

LGTY?

lexelby commented:
Love it, nice elegant solution and it does exactly what I expect it to.

What's up with the change to the line `if aabb > 0 and abab < 0.5 * aabb:`?  Seems like the old and new line do the same thing?  In general I'd love to have some comments in `path_to_curves`, because I have no idea how the math works.

Also, small change, but could you please rename `stitch_curve_even()` to `stitch_curve_evenly()`?

george-steel commented:
> What's up with the change to the line `if aabb > 0 and abab < 0.5 * aabb:`? Seems like the old and new line do the same thing? In general I'd love to have some comments in `path_to_curves`, because I have no idea how the math works.

@lexelby I changed that because I think having the constant on the other side is tidier. I added some comments. 

> Also, small change, but could you please rename `stitch_curve_even()` to `stitch_curve_evenly()`?

done.

lexelby commented:
Okay, I'm still having a hard time understanding your math.  Your equation looks to me to be almost this one:

```
a * b = |a| |b| cos(angle)
```

But you've squared both sides, I think?  Is there some advantage to doing it this way rather than the way I did it in the original algorithm that I'm missing?

george-steel commented:
> Okay, I'm still having a hard time understanding your math. Your equation looks to me to be almost this one:
> 
> ```
> a * b = |a| |b| cos(angle)
> ```
> 
> But you've squared both sides, I think?

@lexelby It's a hand-optimized version of checking if `cos(angle(a,b)) <= sqrt(0.5)` and is defined (I just added a comment to this effect). It's derived as follows:

```
cos(angle(a,b))  <=  sqrt(0.5)
cos(angle(a,b)) * |a| * |b| <=  sqrt(0.5) * |a| * |b|
<a|b>  <=  sqrt(0.5) * |a| * |b|
# apply \x -> x * |x| (which is monotonic) to both sides
<a|b> * |<a|b>|  <=  (sqrt(0.5) * |a| * |b|) * |(sqrt(0.5) * |a| * |b|)|
<a|b> * |<a|b>|  <=  0.5 * <a|a> * <b|b>
abab  <=  0.5 * aabb
```

> Is there some advantage to doing it this way rather than the way I did it in the original algorithm that I'm missing?

There was no corner detection in the original algorithm.

george-steel commented:
@lexelby Is this ready to merge?

lexelby commented:
Ah!  I get it now, thanks.

> There was no corner detection in the original algorithm.

Right you are!  There was in the _original_ original algorithm: 

https://github.com/inkstitch/inkstitch/blob/3b3f890c1256bd7da89da33cc2547d7e8123979c/lib/stitches/running_stitch.py#L49

That comment lies; it's actually checking if the angle is sharper than 60 degrees (cos(60) = 0.5).

That code is taking advantage of the fact that we already needed unit vectors, so we already had them available.  The dot product of unit vectors is just the cosine of the angle between them.

We can do something similar by either converting a and b to unit vectors before dotting them, or dividing the dot product by the magnitude of a and b afterward.  I think that would save several multiplies, since we would only have to do one dot product.  It's like this:

```
<a | b> = |a| |b| cos(angle)
<a | b> / (|a| * |b|) = cos(angle)
```

so we can test for 45 degrees:

```
cos(angle) <= sqrt(0.5)
<a | b> / (|a| * |b|) <= sqrt(0.5)
```

And we can just hardcode sqrt(0.5).

It might be faster to use numpy though.  We're dealing with a list of Point objects, so there's going to be a performance penalty to convert them to a numpy array.  But if we do that, we can get numpy to give us the angle between each successive segment extremely quickly:

```
import numpy as np

points_array = np.array([p.as_tuple() for p in points])

# subtract each successive pair of points
segments = np.diff(points_array)

x_and_y_values = segments.T  # transpose
# x_and_y_values is now [[x1, x2, x3, ...], [y1, y2, y3, ...]]

# get each segment's direction
segment_angles = np.arctan2(x_and_y_values[1], x_and_y_values[0])

# subtract each successive angle to get angles between
angles_between = np.diff(segment_angles)
```

I'm not sure if it would be faster than the dot-product cosine method (though I think it is).  However, it has another significant advantage: it handles angles more correctly.  I can't remember the exact details, but I think the cosine method can miss certain situations like a path that almost doubles back on itself.

george-steel commented:
> We can do something similar by either converting a and b to unit vectors before dotting them, or dividing the dot product by the magnitude of a and b afterward. I think that would save several multiplies, since we would only have to do one dot product. It's like this:

Creating unit vectors requires sqrt, which is slower than an extra multiplication

> It might be faster to use numpy though. We're dealing with a list of Point objects, so there's going to be a performance penalty to convert them to a numpy array. But if we do that, we can get numpy to give us the angle between each successive segment extremely quickly:

Unfortunately vectorizing isn't easy doe to the rejection of duplicate points, and python isn't that bad at arithmetic.

> I'm not sure if it would be faster than the dot-product cosine method (though I think it is). However, it has another significant advantage: it handles angles more correctly. I can't remember the exact details, but I think the cosine method can miss certain situations like a path that almost doubles back on itself.

@lexelby Catching paths that double back on themselves is why I used the cosine method with a monotonic square (using `abs`) instead of a regular square. Your test case earlier showed that.

lexelby commented:
Ah, thanks for explaining all that.  I hadn't noticed the duplicate point filtering.

Great, then your method it is.  Perhaps we should at least store `a * b` and `b * b` in variables, since each is computed twice?  Or are you expecting that Python will optimize that away?

lexelby commented:
Also, just a side note, it's great having you working on the project, thanks for all your hard work on this and the other PRs!  I found your random number stream system really easy to work with in #1803.

george-steel commented:
> Great, then your method it is. Perhaps we should at least store a * b and b * b in variables, since each is computed twice? Or are you expecting that Python will optimize that away?

@lexelby It's such a simple computation (only a few instructions) that even if it isn't optimized, it takes very little time. I decided that the readability of not caching them was better.

I think that if we really want to go for fast speed, it would probably be better to rewrite the core stitches in Rust rather than to try to go for extremely messy, hand-optimized python.

lexelby commented:
Alright, fair point, but I'm really not up for trying to figure out how to do rust modules in the build/packaging system ;)
